EPIC and CASA specification

mawyss opened the pull request:
This pull request contains the EPIC specification, which is based on the discussion from [1].
With this, two new path types are introduced:
- EPIC-HP (EPIC for Hidden Paths)
- EPIC-SAPV (EPIC Source Authentication and Path Validation)

In addition, we propose CASA (Cryptographic Agility for SCION ASes), which allows different entities (ASes, hosts) to agree on the same cryptographic algorithms like PRFs and MACs.

[1] https://docs.google.com/document/d/1n7Gi-WXMM1RzbCS-e3FLdyKNdfI8_mI7I9ndSwjgDpM/edit?usp=sharing

mawyss committed:
First draft of EPIC header

mawyss committed:
Increase precision of timestamp, improve explanation of specification

mawyss committed:
Restructure EPIC header with separate path types

mawyss committed:
Introduce CASA (Cryptographic Agility for SCION ASes)

mawyss committed:
Small correction

mlegner commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,93 @@
+Specification
+=============
+
+Description
```
I like the general idea. One part that is currently missing in the description is the _asymmetry_ that is required for this to work: For all these cryptographic algorithms there is one side with higher efficiency requirements and this side chooses the algorithm. This is essentially mirroring the asymmetry in DRKey.

mawyss replied:
I added a short description of this asymmetry.

mlegner replied:
:+1:

mlegner commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,93 @@
+
+Description
+-----------
+In CASA, every AS chooses exactly **one algorithm** per algorithm 
```
Will these categories and available algorithms also be documented in this file? If not, where will these be specified?

mawyss replied:
From my point of view it would make sense to have them added directly to this document.
I added an additional section to address the concrete algorithms and categories in the future.

mlegner replied:
:+1:

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+
+Path Type: EPIC-HP
+==================
+The EPIC-HP (EPIC for Hidden Paths) header provides improved path authorization for the last hop of the path. 
```
General question about this: Is it enough to add protection to the last hop? If only the last hop gets extra protection, the traffic can only be filtered at the destination AS. Ideally, illegitimate traffic could already be filtered upstream, in this case we would need multiple extra validation fields.

mawyss replied:
I designed EPIC-HP this way based on the discussion with Anapaya (document: "Integrating EPIC into SCION").
To perform an attack on hidden paths, an attacker may still need to also online brute-force some of the other (6-byte) SCION MACs on the path. He may already know (legitimately) some of the hop authenticators (e.g. for the up-segment and parts of the down-segment), for which EPIC L1 would not help anyway. The EPIC L1 per-packet authenticators would only help to increase the security at the remaining hops (that still need to be brute-forced), right? 
With 6-byte MACs as in standard SCION such attacks become extremely hard - or am I missing something?

mlegner replied:
Typically, in a hidden segment, also the second-to-last hop (SLH) would not be publicly announced. If only the last hop were secured by EPIC L1, an on-path adversary could still observe the MAC of the SLH and use it to overload the hidden access link (as packets would only be filtered at the leaf AS). By also securing the SLH with EPIC, this is no longer possible.

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+
+The EPIC-HP header has the following structure:
+   - A *PacketTimestamp* field (8 bytes)
+   - The standard SCION header
```
```suggestion
   - The path header for the standard SCION Path Type
```

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+   - A 4-byte *LHVF* (Last Hop Verification Field) 
+
+The EPIC-HP contains the full SCION header, and also the calculation of the MAC is identical. This allows 
+the destination host to directly send back a SCION answer packet to the source by inverting the path.
```
Can you elaborate on why normal SCION can be used for the reverse direction and what configuration/checks at routers are necessary for that?

mawyss replied:
The content of the fields of an EPIC-HP packet is a superset of the content of the fields of a SCION packet.
If a SCION packet can be reversed, also an EPIC-HP packet can be used to form a reverse SCION packet.
As far as I am aware, reversing SCION packets is already implemented?

mawyss replied:
Maybe for clarification:
As of now, EPIC-HP has the exact same MAC fields as standard SCION, and the field contents are the same.
Only with EPIC-SAPV, the MAC field calculations (and beta calculations) are different.

mlegner replied:
Sorry, my comment was not entirely clear. What I wanted to suggest is adding 1-2 sentences explaining why it is OK (from a security/availability perspective) to allow standard SCION traffic in the reverse direction. And that/why it is necessary that the router enforces the use of some path types in the data plane.

mawyss replied:
I now added some explanation on this.
Let me know if it does not correspond to what you had in mind.

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+    \end{align}
+
+TsRel has a precision of :math:`\text{20 \mu s}` and covers almost 
+one day. When sending packets at high speeds (more than one packet 
```
Does it cover the maximum possible expiration time of hop fields?

mawyss replied:
Based on the 'Hop Field' section, I think that the maximum possible expiration time (ExpTime) is one day.
With 20 microsecond precision, it is not possible to cover the last ~9 minutes.
I reduced the precision to 21 microseconds, now the timestamp covers at least one day.

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+
+
+
+
```
You could remove some of the empty lines and use a consistent number of empty lines throughout the document.

mawyss replied:
Thank you, I removed redundant empty lines.

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+
+.. -------------------------------------------------------------------
+
+Path Type: EPIC-HP
```
You may need to modify line 55 with the newly proposed Path Types.

mawyss replied:
Thank you, I updated that line.

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+Procedures
+----------
+
+**Control plane:**
```
How will the supported/allowed path types be configured/communicated? For example, for a hidden path the standard SCION path type should not be allowed (at least in down direction).

mlegner replied:
Ah, this is described further down, maybe already provide a forward reference?

mawyss replied:
I added a forward reference.
However, if two ASes want to communicate and use each others hidden path, they could not use SCION answer packets, if they both block incoming SCION traffic. But of course they can just communicate using EPIC-HP only. 
Maybe it makes sense to introduce an additional field, that denotes whether the answer can be a reflected SCION packet, to cover this case?

mlegner replied:
That might be a good idea as such a setup (where the two communicating hosts are both between hidden access links) might be quite common. I think a single-bit flag is not too much overhead.
Does anybody else have an opinion on this?

mawyss replied:
I tentatively added the flag to the Path Meta Header for EPIC-HP now.

mlegner commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,425 @@ Because of its special structure, no PathMeta header is needed. There is only a
+Network operators should be able to clearly define which kind of 
+traffic (SCION, EPIC-HP, EPIC-SAPV, and other protocols) they want to
+allow. 
+Therefore, for each AS and every interface pair, an AS can be 
```
The order of the interface pair defines the direction, correct? That means that it is possible to allow SCION traffic from IF 1 to IF 2 but only EPIC from IF 2 to IF 1, correct?

mawyss replied:
Yes, exactly. I added some further description to make it more clear now.

mlegner reviewed (commented):
I like the specification overall, left a number of comments throughout the files.

benrothen commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,93 @@
+that the data plane packets need to introduce an additional field 
+for each AS, which identifies the selected algorithm. For DRKey and 
+EPIC, border routers would need to potentially support many 
+different algorithms, which hinders scalability and performance.
```
also: BRs are likely less agile in adopting new cryptographic algorithms

mawyss replied:
Thank you, I extended the text with that argument.

benrothen commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,93 @@
+In CASA, every AS chooses exactly **one algorithm** per algorithm 
+category. Those categories are globally defined and can either be 
+general (PRF, MAC, ...) or protocol-specific (DRKEY-PRF, 
+EPIC-MAC-HVF, EPIC-MAC-DVF, ...).
```
I assume that in case a protocol-specific algorithm is proposed, it would have priority over the general algorithms?

mawyss replied:
Yes, exactly.

mawyss replied:
I made this fact more visible in the text now.

benrothen commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,93 @@
+entities.
+On the other hand, the protocol-specific categories serve the needs 
+of protocols that have additional requirements like high 
+performance. 
```
What happens if an AS supports the general `PRF` algorithm, but not the `DRKEY-PRF`? Will the protocol then fail or fallback to `PRF`?

mlegner replied:
I would assume that it fails as there is no way to fall back on another algorithm in the data plane.

mawyss replied:
Yes, the protocol will fail in this case, because CASA is based on the assumption that the ASes / hosts support the algorithms from the categories.

benrothen commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,93 @@
+
+Those algorithms are then primarily announced in the beacons, but 
+can also be exchanged in protocol-specific channels, for example 
+when ASes fetch each others first-level DRKeys.
```
This might lead to a lot of redundant information

mawyss replied:
Yes, you are right. 
As of now I changed this part so that the information is only distributed in the beacons.
Therefore it is not necessary for each protocol to individually fetch the information from every on-path AS.

benrothen commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,93 @@
+Advantages
+----------
+With CASA, every AS is free to choose which cryptographic algorithms 
+it wants to support, there is no global agreement necessary. 
```
except for the algorithm categories

mlegner replied:
... and the available options for the algorithms.

mawyss replied:
Thank you, I added those points to the text.

benrothen reviewed (commented):
I left some comments about CASA. I think there are some cases we need to work out: what happens if ASes don't support an algorithm? What happens if an AS changes its algorithm (e.g., resending of beacons)?

mawyss committed:
Address comments, state important parts more precisely

mawyss committed:
Be more precise about SCION path header

Co-authored-by: Markus Legner <mlegner@users.noreply.github.com>

mawyss committed:
Address comments, add reference to CASA

mawyss committed:
Be precise about header

mawyss committed:
Merge branch 's1' of https://github.com/mawyss/scion into s1

mlegner commented on doc/protocols/casa.rst:
```diff
@@ -0,0 +1,120 @@
+EPIC, border routers would need to potentially support many 
+different algorithms, which hinders scalability and performance.
+Also, border routers are likely less agile in adopting new 
+cryptographic algorithms.
```
```suggestion
cryptographic algorithms as they often contain dedicated hardware for packet 
processing.
```

mlegner reviewed (commented):
Thanks for the changes, most of my comments have been resolved. There are some open points with detailed comments.

mawyss committed:
Add reason for lower agility on border routers

Co-authored-by: Markus Legner <mlegner@users.noreply.github.com>

mawyss committed:
Address comments, introduce flag to allow/deny SCION response packets for EPIC-HP

mawyss committed:
Address points discussed with Adrian and Markus:
Protect last link instead of last hop, simplify MAC-chaining, flag for EPIC V3 response packets, simplify EPIC versions, authenticate hash of EPIC authenticators.

mawyss committed:
Drop idea of unauthenticated information in the beacons

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+   - The path header for the standard SCION Path Type, where one bit 
+     of the Path Meta Header is used to indicate whether the sender 
+     accepts SCION response packets.
+   - A 4-byte *SHVF* (Second to last Hop Validation Field)  and a 
```
What about "penultimate" instead of "second to last"?

mawyss replied:
Changed SHVF to PHVF (Penultimate Hop Validation Field).

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+   - A 4-byte *SHVF* (Second to last Hop Validation Field)  and a 
+     4-byte *LHVF* (Last Hop Validation Field)
+
+The EPIC-HP header contains the full SCION header, and also the 
```
SCION path header. The calculation of the hopfield MAC is identical.

mawyss replied:
Adopted this suggestion.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+
+The EPIC-HP header contains the full SCION header, and also the 
+calculation of the MAC is identical. This allows the destination 
+host to directly send back a SCION answer packet to the source by 
```
This enables a destination host to reverse the EPIC-HP header to communicate back to the sender.

mawyss replied:
The idea is that the receiver will not send back EPIC-HP response packets, but a SCION path type response packets.
I added a paragraph to make this more clear now.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+calculation of the MAC is identical. This allows the destination 
+host to directly send back a SCION answer packet to the source by 
+inverting the path. This is allowed from a security perspective, 
+because the SCION answer packets do not leak information that would 
```
It's not obvious that this is the case. You should add a sentence about the why.

mawyss replied:
Added more information on that.
The point is that the SCION path type response packets contain only a subset of the information contained in the EPIC-HP packet.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+inverting the path. This is allowed from a security perspective, 
+because the SCION answer packets do not leak information that would 
+allow unauthorized entities to use the hidden path. To protect the 
+services behind the hidden path from DoS-attacks (only authorized 
```
Preventing DoS-attacks is only one benefit of hidden paths. Another would be to prevent vulnerability scanners to reach services in that AS. In any case, I don't think this last part should be part of this spec. There should be an accompanying design document that explains these aspects in much more detail.

mawyss replied:
Thank you for the note, I will create a design document and move this part in there.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+SCION-Response (S)
+  Indicates whether the sender accepts SCION response packets. A 
```
Unclear what a SCION response packet is.

mawyss replied:
Changed to "SCION path type response packet" (a SCION path type packet where the path is reversed).

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+SCION-Response (S)
+  Indicates whether the sender accepts SCION response packets. A 
+  sender that is not behind a hidden path can set this flag so that 
+  the service knows it has to answer with SCION traffic. A sender 
```
service? rather more generally a receiver.

mawyss replied:
Adopted this suggestion.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+SCION-Response (S)
+  Indicates whether the sender accepts SCION response packets. A 
+  sender that is not behind a hidden path can set this flag so that 
+  the service knows it has to answer with SCION traffic. A sender 
```
What do mean by SCION traffic? Is traffic containing an EPIC-HP path not considered SCION traffic? This needs to be precisely specified.

mawyss replied:
Clarified this in the document:
... it has to answer with SCION path type packets.
With "SCION traffic" I meant SCION path type packets.
In the document, I now use the terms "EPIC-HP packets" and "SCION path type packets" to avoid confusion.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+  Indicates whether the sender accepts SCION response packets. A 
+  sender that is not behind a hidden path can set this flag so that 
+  the service knows it has to answer with SCION traffic. A sender 
+  that is protected by a hidden path itself does not set this flag, 
```
This is an assumption and at most a recommendation. A spec focus on the semantics of what is specified. If you want to have a best practices section, add it somewhere else (e.g., in the design doc or some operational manual)

mawyss replied:
This was not meant as a recommendation, I tried to make it more clear now.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+header, this allows other components (like the replay suppression 
+system) to access it without having to go through any parsing 
+overhead. To achieve an even higher precision of the timestamp, the 
+source is free to allocate additional bits from the PckId to TsRel 
```
I assume each router on the path has to parse the timestamp? Or even if it is just the destination, then how does it now many bits are used to encode the timestamp? This won't work in practice without some way of communicating the format.

mawyss replied:
Thank you for pointing this out. I removed this part, as it would probably introduce more overhead than advantages.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+authenticators and the remaining 10 bytes of second to last and the 
+last authenticator, from the path server. It copies the short 
+authenticators to the corresponding MAC-subfield of the Hop Fields 
+as in standard SCION and adds the current Packet Timestamp. In 
```
SCION path type

mawyss replied:
Adapted.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+    \end{align}
+
+The border routers of the on-path ASes validate and forward the 
+data plane packets as in standard SCION (recalculate 
```
as for the SCION path type

mawyss replied:
Done.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+data plane packets as in standard SCION (recalculate 
+:math:`\sigma_{i}` and compare to the MAC field in the packet). In 
+addition, the second to last hop of the last segment recomputes and 
+verifies the SHVF field (:math:`\sigma_{\text{SH}} = \sigma_{n-1}`, 
```
Why is \sigma_SH == \sigma_{n-1}?

mawyss replied:
I used the notation that the last hop/AS has index n.
Now I changed the index to start at 0 and end at n-1, to make it consistent with the description in the SCION path type section.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+   - **EV = 0:** *unused*
+   - **EV = 1:** *unused (in the future this may be used for EPIC-HP 
+     with improved path authorization on every AS on the path, which 
+     corresponds to EPIC Level 1 in the paper.)*
```
What paper? You need to add that as a reference if you refer to it. I'd rather you would include the relevant bits here such that this part is as stand-alone as possible.

mawyss replied:
Removed the hint to the EPIC paper.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+     future, which corresponds to EPIC Level 3 in the Paper.)*
+
+EPIC Reflected Response (R):
+  Unused at the moment. Meant to indicate that the packet is 
```
Why is it here then?

mawyss replied:
It will be used for the higher EPIC Versions in EPIC-SAPV, once their specification is complete.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+^^^^^^^^^
+We reduce the size of the MAC field to 2 bytes and assign a 4-byte 
+Hop Validation Field (HVF) to the freed space. The total size of the 
+Hop Field stays the same (12 bytes). Note that only two bytes of the 
```
Note,

mawyss replied:
https://dict.leo.org/englisch-deutsch/note%20that

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+Hop Field stays the same (12 bytes). Note that only two bytes of the 
+MAC are involved in the MAC-chaining process (see `Hop Field MAC 
+Computation`_ and `Path Calculation`_), so ASes in EPIC-SAPV can 
+still verifiy the MAC as in standard SCION.
```
standard SCION? Also, this needs to be expanded. It's not clear what it means.

mawyss replied:
I used standard SCION to refer to the SCION path type. Should be more clear now.

mawyss replied:
Explained the procedures in more detail and added example.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+Hop Field
+^^^^^^^^^
+We reduce the size of the MAC field to 2 bytes and assign a 4-byte 
+Hop Validation Field (HVF) to the freed space. The total size of the 
```
Do these work the same way as in EPIC-HP? If yes, mention it. Otherwise, define how they work.

mawyss replied:
They are not calculated the same way as in EPIC-HP.
The computation is specified later in the "Data plane (EPIC Version 2)" section.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+The 16-byte Destination Validation Field is only present if the EPIC 
+Version is 1. The DVF contains the MAC calculated by the source host 
```
Above you mention that EPIC version 1 is unused. Why include this?

mawyss replied:
Thank you for pointing this out, for EPIC Version 1 the DVF is not present, but for EPIC Version 2 it is.

shitz commented on doc/protocols/scion-header.rst:
```diff
@@ -408,3 +408,459 @@ Because of its special structure, no PathMeta header is needed. There is only a
+Validation) contains the following parts:
+
+   - An 8-byte Packet Timestamp (same as for EPIC-HP).
+   - A slightly modified SCION header.
```
SCION path type header - here and everywhere else.

mawyss replied:
Done.

shitz reviewed (changes requested):
Reviewed EPIC-HP. Didn't look at CASA and only skimmed EPIC-SAPV. Can focus simply on EPIC-HP?

mawyss committed:
Address feedback, clarify specification
